python/cudf/cudf/core/column/column.py

```diff
@@ -1948,8 +1948,23 @@ def as_column(
                 raise TypeError(
                     f"Cannot convert a {inferred_dtype} of object type"
                 )
-            elif nan_as_null is False and (
-                pd.isna(arbitrary).any()
+            elif inferred_dtype == "boolean":
+                if cudf.get_option("mode.pandas_compatible"):
+                    if not (
+                        dtype is not None
+                        and dtype == np.dtype("bool")
+                        and not pd.isna(arbitrary).any()
+                    ):
+                        raise MixedTypeError(
+                            f"Cannot have mixed values with {inferred_dtype}"
+                        )
+                elif nan_as_null is False and _has_any_nan(arbitrary):
+                    raise MixedTypeError(
+                        f"Cannot have mixed values with {inferred_dtype}"
+                    )
+            elif (
+                nan_as_null is False
+                and _has_any_nan(arbitrary)
                 and inferred_dtype not in ("decimal", "empty")
```

Comment:
Defer the data introspection to run as late as possible.
```suggestion
                and inferred_dtype not in ("decimal", "empty")
                and _has_any_nan(arbitrary)
```